lib/control/control-commands.c

```diff
@@ -111,13 +111,105 @@ control_connection_stop_process(GString *command, gpointer user_data)
   return result;
 }
 
+
+static gboolean
+_reload(GString *command, gpointer user_data, GString **result)
+{
+  MainLoop *main_loop = (MainLoop *) user_data;
+  gboolean reload_initiated = main_loop_reload_config(main_loop);
+
+  if (reload_initiated)
+    {
+      *result = g_string_new("OK Config reload initiated");
+    }
+  else
+    {
+      *result = g_string_new("FAIL Failed to request config reload");
+    }
+
+  return reload_initiated;
+}
+
 static GString *
 control_connection_reload(GString *command, gpointer user_data)
 {
-  GString *result = g_string_new("OK Config reload initiated");
-  MainLoop *main_loop = (MainLoop *) user_data;
+  GString *result = NULL;
+  _reload(command, user_data, &result);
+  return result;
+}
+
+typedef struct _ReloadCondvar
+{
+  GMutex mutex;
+  GCond condvar;
+  gboolean reloaded;
+} ReloadCondvar;
+
+static ReloadCondvar *
+_reload_condvar_new()
+{
+  ReloadCondvar *self = g_new0(ReloadCondvar, 1);
+  g_mutex_init(&self->mutex);
+  g_cond_init(&self->condvar);
+  self->reloaded = FALSE;
+
+  return self;
+}
+
+static void
+_reload_condvar_free(ReloadCondvar *self)
+{
+  g_mutex_clear(&self->mutex);
+  g_cond_clear(&self->condvar);
+  g_free(self);
+}
+
+static void
+_reload_condvar_emit(ReloadCondvar *self)
+{
+  g_mutex_lock(&self->mutex);
+  self->reloaded = TRUE;
+  g_cond_signal(&self->condvar);
+  g_mutex_unlock(&self->mutex);
+}
+
+static void
+_reload_condvar_wait(ReloadCondvar *self)
+{
+  g_mutex_lock(&self->mutex);
+  while (!self->reloaded)
+    g_cond_wait(&self->condvar, &self->mutex);
```

Comment:
This condition is never met in case of an invalid configuration, because `_config_reloaded` is not triggered in that case. `app_post_config_loaded()` is not called in multiple cases, for example, when `cfg_init` fails:

https://github.com/balabit/syslog-ng/blob/3798aa4dc30e6ff60c8db39b07b9944ddb8da6dc/lib/mainloop.c#L227-L259

Or when `cfg_read_config` fails:
https://github.com/balabit/syslog-ng/blob/3798aa4dc30e6ff60c8db39b07b9944ddb8da6dc/lib/mainloop.c#L293-L300

Reproduction:
- start syslog-ng with a correct config
- add a syntax error to the configuration
- execute `sbin/syslog-ng-ctl reload --wait`

As a result, syslog-ng can't be stopped or reloaded peacefully.